metricbeat/module/mongodb/replstatus/info.go

```diff
@@ -43,85 +43,109 @@ type CollSize struct {
-		return nil, fmt.Errorf("could not decode mongodb op log size: %w", err)
-	}
-
-	// get first and last items in the oplog
+	collection := db.Collection(oplogCol)
 	firstTs, lastTs, err := getOpTimestamp(collection)
 	if err != nil {
-		return nil, fmt.Errorf("could not get operation timestamp in op log: %w", err)
+		return nil, fmt.Errorf("could not get operation timestamps from oplog: %w", err)
 	}
 
-	diff := lastTs - firstTs
-
-	return &oplogInfo{
+	info := &oplogInfo{
 		allocated: oplogSize.MaxSize,
 		used:      oplogSize.Size,
 		firstTs:   firstTs,
 		lastTs:    lastTs,
-		diff:      diff,
-	}, nil
+		diff:      lastTs - firstTs,
+	}
+
+	return info, nil
 }
 
+// getOpTimestamp returns the first and last timestamp of the oplog collection.
 func getOpTimestamp(collection *mongo.Collection) (uint32, uint32, error) {
-
-	// Find both first and last timestamps using $min and $max
-	pipeline := bson.A{
-		bson.M{"$group": bson.M{"_id": 1, "minTS": bson.M{"$min": "$ts"}, "maxTS": bson.M{"$max": "$ts"}}},
+	// NOTE(shmsr):
+	//
+	// When you do db.getReplicationInfo() in monogo shell (mongosh), you can see
+	// 	{
+	//		...
+	// 		tFirst: 'Tue Jan 07 2025 22:33:28 GMT+0530 (India Standard Time)',
+	// 		tLast: 'Wed Jan 08 2025 11:45:07 GMT+0530 (India Standard Time)',
+	// 		now: 'Wed Jan 08 2025 11:45:14 GMT+0530 (India Standard Time)'
+	// 	}
+	// i.e., we get tFirst and tLast from oplog.rs which is the first and last
+	// timestamp of the oplog.
+	// Source from the same is here:
+	// 	https://github.com/mongodb/mongo/blob/20cbee37a0ee4d40b35d08b6a34ade81252f86a8/src/mongo/shell/db.js#L863
+	// This is how they calculate tFirst and tLast:
+	// 	https://github.com/mongodb/mongo/blob/20cbee37a0ee4d40b35d08b6a34ade81252f86a8/src/mongo/shell/db.js#L889
+	// So ideally, we will replicate the same logic here:
+	// 	var firstc = ol.find().sort({$natural: 1}).limit(1);
+	// 	var lastc = ol.find().sort({$natural: -1}).limit(1);
+	//
+	// oplog.rs is designed to scanned in natural ($natural) order. So, when
+	// querying without any sort, it will return the first entry in natural order.
+	// When we sort in reverse natural order (i.e., $natural: -1), it will return
+	// the last entry in natural order.
+	//
+	// The queries use optimizations like FindOne() and SetProjection() to limit
+	// the amount of data that needs to be scanned and processed.
+	// Projection is used to limit the fields returned in the query results i.e., we
+	// only need the timestamp (ts) field. FindOne() is used to retrieve a single
+	// document from the collection (limit: 1).
+
+	ctx := context.TODO()
+
+	var firstDoc struct {
+		Timestamp time.Time `bson:"ts"`
 	}
 
-	cursor, err := collection.Aggregate(context.Background(), pipeline)
+	// Get oldest entry using natural order
+	firstOpts := options.
+		FindOne().
+		SetProjection(bson.D{{Key: "ts", Value: 1}})
+	err := collection.FindOne(ctx, bson.D{}, firstOpts).Decode(&firstDoc)
```

Comment:
Shall we explicitly check if FindOne returns no oplog docuement ?

Reply:
If no results are returned, it will throw an error which is handled. So we can avoid explicit checking of a doc.